File: SdkIntegrationTests/Compatibility/CreditCardTests.m
//
//  CreditCardTests.m
//  SdkIntegrationTests
//
//  Created by Eli Thompson on 6/10/20.
//

#import <XCTest/XCTest.h>
#import <CoreSdk/CoreSdk-Swift.h>
#import "SdkIntegrationTests-Swift.h"

@interface CreditCardTests : XCTestCase

@end

@implementation CreditCardTests

- (void)testCanCreateCreditCard {
    CreditCardInfo *info = [[CreditCardInfo alloc]
            initWithFirstName:@"Dolly"
                     lastName:@"Dog"
                       number:SPRHelpers.secureTestCardNumber
            verificationValue:SPRHelpers.secureVerificationValue
                         year:2030
                        month:12
    ];

    id <SPRClient> client = [SPRHelpers createClient];
    SPRSingleTransaction *transaction = [client createPaymentMethodWithCreditCard:info];

    XCTestExpectation *expectation = [self expectationWithDescription:@"can create"];
    [transaction subscribeOnSuccess:^(SPRTransaction *t) {
        XCTAssertNotNil(t.token);

        SPRCreditCardResult *result = t.creditCard;
        XCTAssertNotNil(result.token);
        XCTAssertEqualObjects(result.firstName, @"Dolly");
        XCTAssertEqualObjects(result.lastName, @"Dog");
        XCTAssertEqualObjects(result.lastFourDigits, @"1111");
        XCTAssertEqual(result.year, 2030);
        XCTAssertEqual(result.month, 12);
        [expectation fulfill];
    }                       onError:^(NSError *error) {
        XCTFail();
        [expectation fulfill];
    }];
    [self waitForExpectationsWithTimeout:10.0 handler:nil];
}

@end

File: SdkTests/CreditCardTests.swift
//
// Created by Eli Thompson on 5/4/20.
//

import XCTest
@testable import CoreSdk

class CreditCardInfoTests: XCTestCase {
    func testCanEncode() throws {
        let client = createSpreedlyClient(envKey: "", envSecret: "")
        let creditCard = CreditCardInfo(
                firstName: "Dolly",
                lastName: "Dog",
                number: client.createSecureString(from: "4111111111111111"),
                verificationValue: client.createSecureString(from: "919"),
                year: 2029,
                month: 12
        )

        let json = try creditCard.toJson()

        let expected = try """
                           {
                             "first_name" : "Dolly",
                             "last_name" : "Dog",
                             "month" : 12,
                             "number" : "4111111111111111",
                             "verification_value" : "919",
                             "year" : 2029
                           }
                           """.data(using: .utf8)!.decodeJson()

        XCTAssertEqual(expected as NSObject, json as NSObject)
    }

    func testCanEncodeWithFullName() throws {
        let client = createSpreedlyClient(envKey: "", envSecret: "")
        let creditCard = CreditCardInfo(
                fullName: "Dolly Dog",
                number: client.createSecureString(from: "4111111111111111"),
                verificationValue: client.createSecureString(from: "919"),
                year: 2029,
                month: 12
        )

        let json = try creditCard.toJson()

        let expected = try """
                           {
                             "full_name" : "Dolly Dog",
                             "month" : 12,
                             "number" : "4111111111111111",
                             "verification_value" : "919",
                             "year" : 2029
                           }
                           """.data(using: .utf8)!.decodeJson()

        XCTAssertEqual(expected as NSObject, json as NSObject)
    }
}
